### PR TITLE
Modify presto to use 64bit hashes returned from types.

### PR DESCRIPTION
**Pull request by fiedukow**
This implementation of issue: https://github.com/prestodb/presto/issues/4133
Here are results of benchmarks pointed inside the issue:
Before changes: https://gist.github.com/fiedukow/81ae0e1b31f4e086d3befa75aa7c091f
After changes: https://gist.github.com/fiedukow/03739319f2f10d0fcf2467c53671457f

It seems that this has positive impact on performance for large group by hashing.

Note that due to nature of some parts of code and external dependencies (eg. airlift) 32 bit hashes are still required in some places. There is longHashToShortHash function introduce to cover conversion without changing local presto code back to 32 bits hashes.

@cberner

### CODE REVIEW

**Comment by ilfrin** [comment]
@martint @dain, we discussed this during our last chat about performance improvements

File: presto-main/src/main/java/com/facebook/presto/util/Types.java [hunk 1]
```diff
@@ -32,4 +32,15 @@ private Types() {}
...
         return target.cast(value);
     }
+
+    /**
+     * Uses 64 bit hash and compact its entropy in 32 bits.
+     * @deprecated This function should only be used for keeping backward compatibility with external libraries using 32 hashes.
+     * @param hash - the 64bit (long) hash
+     * @return 32bit (int) hash created based on long hash
+     */
+    public static int longHashToShortHash(long hash)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/util/Types.java]: longHashToIntegerHash?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/BigintType.java [hunk 2]
```diff
@@ -61,10 +61,9 @@ public boolean equalTo(Block leftBlock, int leftPosition, Block rightBlock, int
     }
 
     @Override
-    public int hash(Block block, int position)
+    public long hash(Block block, int position)
     {
-        long value = block.getLong(position, 0);
-        return (int) (value ^ (value >>> 32));
+        return block.getLong(position, 0);
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/BigintType.java]: I think it needs to be hashed even it is a long

File: presto-spi/src/main/java/com/facebook/presto/spi/type/DoubleType.java [hunk 3]
```diff
@@ -60,10 +60,9 @@ public boolean equalTo(Block leftBlock, int leftPosition, Block rightBlock, int
     }
 
     @Override
-    public int hash(Block block, int position)
+    public long hash(Block block, int position)
     {
-        long value = block.getLong(position, 0);
-        return (int) (value ^ (value >>> 32));
+        return block.getLong(position, 0);
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/DoubleType.java]: I think it should be hashed anyway

File: presto-main/src/test/java/com/facebook/presto/block/AbstractTestBlock.java [hunk 4]
```diff
@@ -146,7 +146,6 @@
 
             for (int offset = 0; offset < length - 3; offset++) {
                 assertEquals(block.getSlice(position, offset, 3), expectedSliceValue.slice(offset, 3));
-                assertEquals(block.hash(position, offset, 3), expectedSliceValue.hashCode(offset, 3));
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/block/AbstractTestBlock.java]: why did you remove this?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/BigintType.java (hunk 2 above)
Comment (fiedukow) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/BigintType.java]: I don't think it needs to be hashed. It wasn't before. AFAIU real hashing with nice distribution is done later in the code when it's really needed.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/DoubleType.java (hunk 3 above)
Comment (fiedukow) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/DoubleType.java]: I don't think it needs to be hashed. It wasn't before. AFAIU real hashing with nice distribution is done later in the code when it's really needed.

File: presto-main/src/test/java/com/facebook/presto/block/AbstractTestBlock.java (hunk 4 above)
Comment (fiedukow) [inline presto-main/src/test/java/com/facebook/presto/block/AbstractTestBlock.java]: I don't think anyone relies on that and there is no reason to do this I think.
This is explicitly testing if our Block code is using exactly same hashing strategy as airlift Slice which doesn't make much sense, especially as airlift is using 32bit hash in this scenario.

I probably could replace expectedSliceValue.hashCode with XxHash64.hash(expectedSliceValue) to get equality, but does this really makes any sense to test that? It's like checking if we are still using same implementation.

File: presto-main/src/main/java/com/facebook/presto/operator/BigintGroupByHash.java [hunk 5]
```diff
@@ -303,9 +303,9 @@ private void rehash()
         this.valuesByGroupId.ensureCapacity(maxFill);
     }
 
-    private static int getHashPosition(long rawHash, int mask)
+    private static long getHashPosition(long rawHash, int mask)
     {
-        return ((int) murmurHash3(rawHash)) & mask;
+        return murmurHash3(rawHash) & mask;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/BigintGroupByHash.java]: murmurHash3 produces a 32-bit hash. We should change it to XXHash64.

File: presto-main/src/main/java/com/facebook/presto/operator/MultiChannelGroupByHash.java [hunk 6]
```diff
@@ -397,7 +397,7 @@ private boolean positionEqualsCurrentRow(long address, int hashPosition, int pos
         return hashStrategy.positionEqualsRow(decodeSliceIndex(address), decodePosition(address), position, page, hashChannels);
     }
 
-    private static int getHashPosition(int rawHash, int mask)
+    private static long getHashPosition(long rawHash, int mask)
     {
         return murmurHash3(rawHash) & mask;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/MultiChannelGroupByHash.java]: Change to XXHash64, per comment above.

File: presto-main/src/main/java/com/facebook/presto/operator/ParallelHashBuilder.java [hunk 7]
```diff
@@ -257,8 +257,8 @@ public void addInput(Page page)
             // build a block containing the partition id of each position
             BlockBuilder blockBuilder = BIGINT.createBlockBuilder(new BlockBuilderStatus(), page.getPositionCount());
             for (int position = 0; position < page.getPositionCount(); position++) {
-                int rawHash = hashGenerator.hashPosition(position, page);
-                int partition = murmurHash3(rawHash) & parallelStreamMask;
+                long rawHash = hashGenerator.hashPosition(position, page);
+                int partition = (int) (murmurHash3(rawHash) & parallelStreamMask);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/ParallelHashBuilder.java]: Change to XXHash64

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java [hunk 8]
```diff
@@ -69,9 +69,9 @@ public long getJoinPosition(int position, Page page)
     }
 
     @Override
-    public long getJoinPosition(int position, Page page, int rawHash)
+    public long getJoinPosition(int position, Page page, long rawHash)
     {
-        int partition = murmurHash3(rawHash) & partitionMask;
+        int partition = (int) murmurHash3(rawHash) & partitionMask;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java]: Change to XXHash64

File: presto-main/src/main/java/com/facebook/presto/sql/gen/InCodeGenerator.java [hunk 9]
```diff
@@ -210,7 +211,13 @@ public BytecodeNode generateExpression(Signature signature, BytecodeGeneratorCon
                         .comment("lookupSwitch(hashCode(<stackValue>))")
                         .dup(javaType)
                         .append(invoke(hashCodeBinding, hashCodeSignature))
+                        //.invokeStatic(Types.class, "longHashToShortHash", int.class, long.class)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/InCodeGenerator.java]: Why did you inline the operations from longHashToShortHash?

File: presto-main/src/main/java/com/facebook/presto/util/Types.java [hunk 10]
```diff
@@ -32,4 +32,15 @@ private Types() {}
                 value.getClass().getName());
         return target.cast(value);
     }
+
+    /**
+     * Uses 64 bit hash and compact its entropy in 32 bits.
+     * @deprecated This function should only be used for keeping backward compatibility with external libraries using 32 hashes.
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/util/Types.java]: What libraries are you referring to? Nothing external to Presto should be relying on the shape of hashes.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractFixedWidthBlock.java [hunk 11]
```diff
@@ -113,13 +114,14 @@ public boolean bytesEqual(int position, int offset, Slice otherSlice, int otherO
...
-    public int hash(int position, int offset, int length)
+    public long hash(int position, int offset, int length)
     {
         checkReadablePosition(position);
         if (isNull(position)) {
             return 0;
         }
-        return getRawSlice().hashCode(valueOffset(position) + offset, length);
+
+        return XxHash64.hash(getRawSlice(), valueOffset(position) + offset, length); // TODO: this can be patched into airlift
```
Comment (martint) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractFixedWidthBlock.java]: I'm not sure I understand the TODO

**Comment by martint** [comment]
Some minor changes needed, but overall looks good.

File: presto-main/src/main/java/com/facebook/presto/operator/BigintGroupByHash.java (hunk 5 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/BigintGroupByHash.java]: Are you sure about that?
It looks like it returns 64 bit hashes:
https://ideone.com/ykpZ47 (murmurHash3 copied from decompiled it.unimi.dsi.fastutil.HashCommon class file).

File: presto-main/src/main/java/com/facebook/presto/sql/gen/InCodeGenerator.java (hunk 9 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/InCodeGenerator.java]: I was wondering which version is preferred here as the pattern in code is not clear (some parts are generated to bytecode and some are invoked). If you prefer other solution I don't see any obvious downside of that.

File: presto-main/src/main/java/com/facebook/presto/util/Types.java (hunk 10 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/util/Types.java]: If I understand that correctly, nothing external relies on shape of presto hashes. However, Fastutil Hashset requires hash providers (HashStrategies) and our implementation uses hashing function to provide them. This is why this function is required to provide them without changing unrelated code right now.

Even more tricky point is using 64 bit hashes in InCodeGenerator (switch varian) as java only support switches on integers, without such fallback we will have to redesign whole generator to avoid switches.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractFixedWidthBlock.java (hunk 11 above)
Comment (fiedukow) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractFixedWidthBlock.java]: This TODO is invalid, I forgot to remove it - as you can see it already is available in airlift - I just missed it on first iteration ;-)

**Comment by fiedukow** [comment]
Fixed some comments and rebased to current master (INTEGER integrated).

File: presto-main/src/main/java/com/facebook/presto/operator/BigintGroupByHash.java (hunk 5 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/BigintGroupByHash.java]: That's interesting. Murmur3 is either a 32 bit or a128 bit hash. I've no idea what hash that is :)

It's ok to leave as is, then. Eventually, we should switch to xxhash64 for consistency, anyway.

File: presto-main/src/main/java/com/facebook/presto/util/Types.java (hunk 10 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/util/Types.java]: Ah, ok. The comment was not very clear. Maybe add that explanation to the javadoc.

File: presto-main/src/main/java/com/facebook/presto/util/Types.java [hunk 12]
```diff
@@ -32,4 +32,15 @@ private Types() {}
...
+     * @param hash - the 64bit (long) hash
+     * @return 32bit (int) hash created based on long hash
+     */
+    public static int longHashToIntegerHash(long hash)
+    {
+        return (int) ((hash >>> 32) ^ hash);
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/util/Types.java]: We should use `Long.hashCode(value)` instead of this method everywhere (including the generated bytecode).

File: presto-main/src/main/java/com/facebook/presto/sql/gen/InCodeGenerator.java [hunk 13]
```diff
@@ -210,7 +211,13 @@ public BytecodeNode generateExpression(Signature signature, BytecodeGeneratorCon
                         .comment("lookupSwitch(hashCode(<stackValue>))")
                         .dup(javaType)
                         .append(invoke(hashCodeBinding, hashCodeSignature))
+                        //.invokeStatic(Types.class, "longHashToIntegerHash", int.class, long.class)
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/InCodeGenerator.java]: use `Long.hashCode(value)` instead.

File: presto-main/src/main/java/com/facebook/presto/util/Types.java (hunk 10 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/util/Types.java]: (function removed as @dain suggested)

File: presto-main/src/main/java/com/facebook/presto/util/Types.java (hunk 12 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/util/Types.java]: done

**Comment by martint** [comment]
Merged, thanks!
